implements RFC 463; produce gcc compatible dep files

ringabout opened the pull request:
closes https://github.com/nim-lang/RFCs/issues/463

todo

- [x] changelog
- [x] better error messages
- [x] allow specifying the nim executable 
- [x] allow specifying nimcache
- [x] help messages

ringabout committed:
implements RFC 463; produce gcc compatible dep files

ringabout committed:
typo

ringabout committed:
typo

ringabout committed:
off by one

CyberTailor commented:
Could you [rebase](https://git-rebase.io/) commits for cleaner logs?

CyberTailor commented:
So `nimgendeps` is a wrapper around `nim c`? Very clever!

CyberTailor commented:
Also I'd:

1) pass all unknown options to Nim
2) add "--nim" option to set Nim executable
2) respect "--nimcache" option
3) add help output

CyberTailor commented:
Depfile should be written not for "src" but for out binary.

ringabout committed:
finish tasks (#20230)

* Improve error message for `strutils.addf` (#20157)

Co-authored-by: ringabout <43030857+ringabout@users.noreply.github.com>

* fixes #20153; do not escape `_` for mysql [backport] (#20164)

* fixes #20153; do not escape `_` for mysql

* add a test

* Update db_mysql.nim

* Update tdb_mysql.nim

Co-authored-by: Clay Sweetser <Varriount@users.noreply.github.com>

* [minor] don't find `"Hint: gc"` for action (#20170)

* fixes links in the readme (#20167)

* update the docs of arc following up #19749 (#19752)

Co-authored-by: flywind <43030857+xflywind@users.noreply.github.com>

* fixes broken ssl tests (#20181)

* bootstrap the compiler with nimPreviewSlimSystem (#20176)

* bootstrap the compiler with nimPreviewSlimSystem

* threads

* docs: fix some spelling errors (#19816)

* docs: fix some spelling errors

* contributing: fix spelling error

Co-authored-by: konsumlamm <44230978+konsumlamm@users.noreply.github.com>

* Update contributing.md

* Update intern.md

* Update manual.md

* Update manual_experimental_strictnotnil.md

* Update nimgrep_cmdline.txt

* Update pegdocs.txt

* Update testament.md

* Update tut1.md

Co-authored-by: konsumlamm <44230978+konsumlamm@users.noreply.github.com>
Co-authored-by: Clay Sweetser <Varriount@users.noreply.github.com>

* Build compiler with --noNimblePath (#20168)

- Fixes https://github.com/nim-lang/Nim/issues/18840

* help our poor CI; don't run CI on other branches for push (#20184)

* fix broken runnableExamples for getWeeksInIsoYear (#20193)

Based on what I understand from [Wikipedia](https://en.wikipedia.org/wiki/ISO_week_date#Weeks_per_year), 2001 does not have 53 weeks, but 2004, 2009, 2015, 2020 do. The years 2000 and 2001 seem to be copy pasted from the `getDaysInYear` example above. The result of `getWeeksInIsoYear` also seem to match up with Wikipedia.

That means these runnableExamples were never tested. Why is this the case? I only discovered this in #20091.

* improve deprecation error messages (#20197)

* Show beatutiful html instead of ugly markdown preview (#20196)

* closes #6559; add testcase (#20200)

* CI upgrade to ubuntu-20.04 (#20182)

* CI upgrade to ubuntu-20.04 

The ubuntu-18.04 environment is deprecated, consider switching to ubuntu-20.04(ubuntu-latest), or ubuntu-22.04 instead. For more details see https://github.com/actions/virtual-environments/issues/6002

* Update azure-pipelines.yml

* Markdown code blocks part 4 (#20189)

No logic was added, just 8 more files have been migrated.

* closes #15316; add testcase (#20213)

* add version-1-6 and version-1-2 to triggered branches (#20214)

* add version-1-6 and version-1-2 to triggered branches

* Update .github/workflows/ci_packages.yml

* use quote

* closes #12955; add testcase (#20223)

* add more

Co-authored-by: konsumlamm <44230978+konsumlamm@users.noreply.github.com>
Co-authored-by: Clay Sweetser <Varriount@users.noreply.github.com>
Co-authored-by: flywind <43030857+xflywind@users.noreply.github.com>
Co-authored-by: ee7 <45465154+ee7@users.noreply.github.com>
Co-authored-by: Ivan Yonchovski <yyoncho@users.noreply.github.com>
Co-authored-by: metagn <metagngn@gmail.com>
Co-authored-by: Andrey Makarov <ph.makarov@gmail.com>

ringabout committed:
Merge remote-tracking branch 'upstream/devel' into pr_gendeps

ringabout committed:
Merge branch 'devel' into pr_gendeps

ringabout commented on changelog.md:
```diff
@@ -140,3 +140,8 @@
   replace `pkgs` by `pkgs2`.
 
+
+- There is a new switch `--nimMainPrefix:prefix` to influence the `NimMain` that the
+  compiler produces. This is particularly useful for generating static libraries.
```
```suggestion
```

ringabout committed:
Update changelog.md

ringabout commented on changelog.md:
```diff
@@ -140,3 +140,6 @@
   a simple configuration change setting that you can do yourself too. In `$nim/config/nim.cfg`
   replace `pkgs` by `pkgs2`.
 
+
+
```
```suggestion
```

ringabout committed:
Update changelog.md

ringabout commented:
> 1) pass all unknown options to Nim
> 2) add "--nim" option to set Nim executable
> 3) respect "--nimcache" option
> 4) add help output
> 5) Depfile should be written not for "src" but for out binary.

@CyberTailor Thanks for your suggestions. The points above are solved.

Varriount commented on tools/nimgendeps.nim:
```diff
@@ -0,0 +1,105 @@
+    result.add src[i]
+
+proc writeDepfile(src: string, option: DepOption) =
+  let filename = extractFilename(src)
+  let jsonFile = filename.changeFileExt("json")
+  if option.nimcacheDir.len == 0:
+    option.nimcacheDir = getTempDir() / "nimgccdeps"
+  if option.passNim.len == 0:
+    option.passNim = "c"
+  if option.nimBinary.len == 0:
+    option.nimBinary = "nim"
+  let (msg, exitCode) = execCmdEx(fmt"{option.nimBinary} {option.passNim} --d:nimBetterRun --nimcache:{option.nimcacheDir} {src}")
+  doAssert exitCode == 0, msg
+  let jsonData = parseJson(readFile(option.nimcacheDir / jsonFile))
+  let deps = jsonData["depfiles"]
+  var f: File
+  if option.outFile.len == 0:
+    f = stdout
+  else:
+    f = open(option.outFile, fmWrite)
+  try:
+    let target = src.changeFileExt(when defined(windows): ".exe" else: "").handleMakeEscape()
+    f.write(target & ": \\" & '\n')
+
+    for i in 0 ..< deps.len-1:
+      let path = deps[i][0].getStr.handleMakeEscape()
+      f.write('\t' & path & " \\" & '\n')
+    let path = deps[^1][0].getStr.handleMakeEscape()
+    f.write('\t' & path)
+  finally:
+    f.close()
```
Would this work instead?
```suggestion
  if option.nimcacheDir.len == 0:
    option.nimcacheDir = getTempDir() / "nimgccdeps"
  if option.passNim.len == 0:
    option.passNim = "c"
  if option.nimBinary.len == 0:
    option.nimBinary = "nim"

  let (msg, exitCode) = execCmdEx(
    fmt"{option.nimBinary} {option.passNim} --d:nimBetterRun --nimcache:{option.nimcacheDir} {src}"
  )
  doAssert exitCode == 0, msg
  
  let jsonFileName = changeFileExt(extractFileName(src), ".json")
  let jsonFilePath = option.nimcacheDir / jsonFile
  let jsonFileData = parseJson(readFile(jsonFilePath))
  let depFiles = jsonFileData["depfiles"]

  var extension: string
  if defined(windows):
    extension = ".exe"
  else:
    extension = ""
  
  var outFile: File
  if option.outFile.len == 0:
    outFile = stdout
  else:
    outFile = open(option.outFile, fmWrite)

  try:
    let target = changeFileExt(src, extension)
    outFile.write(handleMakeEscape(target))
    outFile.write(": ")

    for i in 0 ..< depFiles.len:
      let path = getStr(depFiles[i][0])
      outfile.write("\\\n")
      outFile.write('\t')
      outFile.write(handleMakeEscape(path))

    outfile.write("\n")
  finally:
    f.close()
```

Varriount commented on tools/nimgendeps.nim:
```diff
@@ -0,0 +1,105 @@
+  var passNim = ""
+  var outFile = ""
+  var src = ""
+  var nimBinary = ""
+  var nimcacheDir = ""
+  var isStop = false
+  while true:
+    p.next()
+    case p.kind
+    of cmdEnd: break
+    of cmdLongOption:
+      if p.key.cmpIgnoreStyle("passNim") == 0:
+        passNim = p.val
+      elif p.key.cmpIgnoreStyle("out") == 0:
+        outFile = p.val
+      elif p.key.cmpIgnoreStyle("nim") == 0:
+        nimBinary = p.val
+      elif p.key.cmpIgnoreStyle("nimcache") == 0:
+        nimcacheDir = p.val
+      else:
+        echo "unexpected option: ", p.key
+        isStop = true
+
+    of cmdShortOption:
+      if p.key == "o":
+        outFile = p.val
+      else:
+        echo "unexpected option: ", p.key
+        isStop = true
+    of cmdArgument:
+      src = p.key
+      break
+
+  if isStop:
+    discard
+  elif src.len == 0:
+    echo helpText
+  else:
+    writeDepfile(src, DepOption(passNim: passNim, outFile: outFile,
+                  nimBinary: nimBinary, nimcacheDir: nimcacheDir))
```
```suggestion
  var passNim = ""
  var outFile = ""
  var sourceFile = ""
  var nimBinary = ""
  var nimcacheDir = ""
  var argumentError = false

  let parser = initOptParser()
  while true:
    parser.next()

    case parser.kind
    of cmdEnd:
      break

    of cmdLongOption:
      if cmpIgnoreStyle(parser.key, "passNim") == 0:
        passNim = parser.val
      elif cmpIgnoreStyle(parser.key, "out") == 0:
        outFile = parser.val
      elif cmpIgnoreStyle(parser.key, "nim") == 0:
        nimBinary = parser.val
      elif cmpIgnoreStyle(parser.key, "nimcache") == 0:
        nimcacheDir = parser.val
      else:
        echo "unexpected option: ", parser.key
        argumentError = true

    of cmdShortOption:
      if parser.key == "o":
        outFile = parser.val
      else:
        echo "unexpected option: ", parser.key
        argumentError = true

    of cmdArgument:
      if sourceFile.len == 0:
        sourceFile = parser.key
      else:
        echo "unexpected argument: ", parser.key
        argumentError = true

  if argumentError:
    discard

  elif sourceFile.len == 0:
    echo helpText
  else:
    writeDepfile(
      src,
      DepOption(
        passNim: passNim,
        outFile: outFile,
        nimBinary: nimBinary,
        nimcacheDir: nimcacheDir
      )
    )
```

Varriount commented:
@ringabout Please see comments
